File: v1/form_builder.go
package smartform

import (
	"strings"

	"github.com/juicycleff/smartform/v1/template"
)

// FormBuilder provides a fluent API for creating form schemas
type FormBuilder struct {
	schema *FormSchema
}

// NewForm creates a new form builder
func NewForm(id, title string) *FormBuilder {
	return &FormBuilder{
		schema: NewFormSchema(id, title),
	}
}

// NewAuthForm creates a new authentication form builder
func NewAuthForm(id, title string, authType AuthStrategy) *FormBuilder {
	return &FormBuilder{
		schema: NewAuthFormSchema(id, title, authType),
	}
}

// Description sets the form description
func (fb *FormBuilder) Description(description string) *FormBuilder {
	fb.schema.Description = description
	return fb
}

// FormType sets the form type
func (fb *FormBuilder) FormType(formType FormType) *FormBuilder {
	fb.schema.Type = formType
	return fb
}

// AuthType sets the authentication type for auth forms
func (fb *FormBuilder) AuthType(authType AuthStrategy) *FormBuilder {
	if fb.schema.Type == FormTypeAuth {
		fb.schema.AuthType = authType
	}
	return fb
}

// Property sets a custom property on the form
func (fb *FormBuilder) Property(key string, value interface{}) *FormBuilder {
	fb.schema.Properties[key] = value
	return fb
}

// AddField adds a field to the form
func (fb *FormBuilder) AddField(field *Field) *FormBuilder {
	fb.schema.Fields = append(fb.schema.Fields, field)
	return fb
}

// AddFields adds multiple fields to the form
func (fb *FormBuilder) AddFields(fields ...*Field) *FormBuilder {
	fb.schema.Fields = append(fb.schema.Fields, fields...)
	return fb
}

// TextField adds a text field to the form
func (fb *FormBuilder) TextField(id, label string) *FieldBuilder {
	field := NewFieldBuilder(id, FieldTypeText, label)
	fb.AddField(field.Build())
	return field
}

// TextareaField adds a textarea field to the form
func (fb *FormBuilder) TextareaField(id, label string) *FieldBuilder {
	field := NewFieldBuilder(id, FieldTypeTextarea, label)
	fb.AddField(field.Build())
	return field
}

// NumberField adds a number field to the form
func (fb *FormBuilder) NumberField(id, label string) *FieldBuilder {
	field := NewFieldBuilder(id, FieldTypeNumber, label)
	fb.AddField(field.Build())
	return field
}

// EmailField adds an email field to the form
func (fb *FormBuilder) EmailField(id, label string) *FieldBuilder {
	field := NewFieldBuilder(id, FieldTypeEmail, label)
	fb.AddField(field.Build())
	return field
}

// PasswordField adds a password field to the form
func (fb *FormBuilder) PasswordField(id, label string) *FieldBuilder {
	field := NewFieldBuilder(id, FieldTypePassword, label)
	fb.AddField(field.Build())
	return field
}

// SelectField adds a select field to the form
func (fb *FormBuilder) SelectField(id, label string) *FieldBuilder {
	field := NewFieldBuilder(id, FieldTypeSelect, label)
	fb.AddField(field.Build())
	return field
}

// MultiSelectField adds a multi-select field to the form
func (fb *FormBuilder) MultiSelectField(id, label string) *FieldBuilder {
	field := NewFieldBuilder(id, FieldTypeMultiSelect, label)
	fb.AddField(field.Build())
	return field
}

// CheckboxField adds a checkbox field to the form
func (fb *FormBuilder) CheckboxField(id, label string) *FieldBuilder {
	field := NewFieldBuilder(id, FieldTypeCheckbox, label)
	fb.AddField(field.Build())
	return field
}

// RadioField adds a radio button field to the form
func (fb *FormBuilder) RadioField(id, label string) *FieldBuilder {
	field := NewFieldBuilder(id, FieldTypeRadio, label)
	fb.AddField(field.Build())
	return field
}

// DateField adds a date field to the form
func (fb *FormBuilder) DateField(id, label string) *FieldBuilder {
	field := NewFieldBuilder(id, FieldTypeDate, label)
	fb.AddField(field.Build())
	return field
}

// TimeField adds a time field to the form
func (fb *FormBuilder) TimeField(id, label string) *FieldBuilder {
	field := NewFieldBuilder(id, FieldTypeTime, label)
	fb.AddField(field.Build())
	return field
}

// DateTimeField adds a datetime field to the form
func (fb *FormBuilder) DateTimeField(id, label string) *FieldBuilder {
	field := NewFieldBuilder(id, FieldTypeDateTime, label)
	fb.AddField(field.Build())
	return field
}

// FileField adds a file upload field to the form
func (fb *FormBuilder) FileField(id, label string) *FieldBuilder {
	field := NewFieldBuilder(id, FieldTypeFile, label)
	fb.AddField(field.Build())
	return field
}

// ImageField adds an image upload field to the form
func (fb *FormBuilder) ImageField(id, label string) *FieldBuilder {
	field := NewFieldBuilder(id, FieldTypeImage, label)
	fb.AddField(field.Build())
	return field
}

// SwitchField adds a switch field to the form
func (fb *FormBuilder) SwitchField(id, label string) *FieldBuilder {
	field := NewFieldBuilder(id, FieldTypeSwitch, label)
	fb.AddField(field.Build())
	return field
}

// SliderField adds a slider field to the form
func (fb *FormBuilder) SliderField(id, label string) *FieldBuilder {
	field := NewFieldBuilder(id, FieldTypeSlider, label)
	fb.AddField(field.Build())
	return field
}

// RatingField adds a rating field to the form
func (fb *FormBuilder) RatingField(id, label string) *FieldBuilder {
	field := NewFieldBuilder(id, FieldTypeRating, label)
	fb.AddField(field.Build())
	return field
}

// ColorField adds a color picker field to the form
func (fb *FormBuilder) ColorField(id, label string) *FieldBuilder {
	field := NewFieldBuilder(id, FieldTypeColor, label)
	fb.AddField(field.Build())
	return field
}

// HiddenField adds a hidden field to the form
func (fb *FormBuilder) HiddenField(id string, value interface{}) *FieldBuilder {
	field := NewFieldBuilder(id, FieldTypeHidden, "")
	field.DefaultValue(value)
	fb.AddField(field.Build())
	return field
}

// RichTextField adds a rich text editor field to the form
func (fb *FormBuilder) RichTextField(id, label string) *FieldBuilder {
	field := NewFieldBuilder(id, FieldTypeRichText, label)
	fb.AddField(field.Build())
	return field
}

// SectionField adds a section separator to the form
func (fb *FormBuilder) SectionField(id, label string) *FieldBuilder {
	field := NewFieldBuilder(id, FieldTypeSection, label)
	fb.AddField(field.Build())
	return field
}

// GroupField adds a group field to the form
func (fb *FormBuilder) GroupField(id, label string) *GroupFieldBuilder {
	field := NewGroupFieldBuilder(id, label)
	fb.AddField(field.Build())
	return field
}

// ArrayField adds an array field to the form
func (fb *FormBuilder) ArrayField(id, label string) *ArrayFieldBuilder {
	field := NewArrayFieldBuilder(id, label)
	fb.AddField(field.Build())
	return field
}

// OneOfField adds a oneOf field to the form
func (fb *FormBuilder) OneOfField(id, label string) *OneOfFieldBuilder {
	field := NewOneOfFieldBuilder(id, label)
	fb.AddField(field.Build())
	return field
}

// AnyOfField adds an anyOf field to the form
func (fb *FormBuilder) AnyOfField(id, label string) *AnyOfFieldBuilder {
	field := NewAnyOfFieldBuilder(id, label)
	fb.AddField(field.Build())
	return field
}

// APIField adds an API integration field to the form
func (fb *FormBuilder) APIField(id, label string) *APIFieldBuilder {
	field := NewAPIFieldBuilder(id, label)
	fb.AddField(field.Build())
	return field
}

// AuthField adds an authentication field to the form
func (fb *FormBuilder) AuthField(id, label string) *AuthFieldBuilder {
	field := NewAuthFieldBuilder(id, label)
	fb.AddField(field.Build())
	return field
}

// OAuthField adds an OAuth authentication field to the form
func (fb *FormBuilder) OAuthField(id, label string) *OAuth2Builder {
	if id == "" {
		id = "oauth"
	}

	field := NewOAuth2Builder(id, label)
	fb.AddField(field.Build())
	return field
}

// BasicAuthField adds a basic authentication field to the form
func (fb *FormBuilder) BasicAuthField(id, label string) *BasicAuthBuilder {
	if id == "" {
		id = "basic"
	}
	field := NewBasicAuthBuilder(id, label)
	fb.AddField(field.Build())
	return field
}

// APIKeyField adds an API key authentication field to the form
func (fb *FormBuilder) APIKeyField(id, label string) *APIKeyBuilder {
	if id == "" {
		id = "key"
	}

	field := NewAPIKeyBuilder(id, label)
	fb.AddField(field.Build())
	return field
}

// JWTField adds a JWT authentication field to the form
func (fb *FormBuilder) JWTField(id, label string) *JWTBuilder {
	if id == "" {
		id = "jwt"
	}
	field := NewJWTBuilder(id, label)
	fb.AddField(field.Build())
	return field
}

// SAMLField adds a SAML authentication field to the form
func (fb *FormBuilder) SAMLField(id, label string) *SAMLBuilder {
	if id == "" {
		id = "saml"
	}
	field := NewSAMLBuilder(id, label)
	fb.AddField(field.Build())
	return field
}

// BranchField adds a workflow branch field to the form
func (fb *FormBuilder) BranchField(id, label string) *BranchFieldBuilder {
	field := NewBranchFieldBuilder(id, label)
	fb.AddField(field.Build())
	return field
}

// CustomField Custom adds a custom field to the form
func (fb *FormBuilder) CustomField(id, label string) *CustomFieldBuilder {
	if id == "" {
		id = "custom"
	}
	field := NewCustomFieldBuilder(id, label)
	fb.AddField(field.Build())
	return field
}

// RegisterVariable registers a variable in the form
func (fb *FormBuilder) RegisterVariable(name string, value interface{}) *FormBuilder {
	fb.schema.RegisterVariable(name, value)
	return fb
}

// RegisterVariableFunction registers a function in the form
func (fb *FormBuilder) RegisterVariableFunction(name string, fn template.TemplateFunction) *FormBuilder {
	fb.schema.RegisterVariableFunction(name, fn)
	return fb
}

// Build finalizes and returns the form schema
func (fb *FormBuilder) Build() *FormSchema {
	fb.registerDynamicFunctions()

	return fb.schema
}

func (fb *FormBuilder) registerDynamicFunctions() {
	for _, field := range fb.schema.Fields {
		fb.registerFieldDynamicFunctions(field, "")
	}
}

// Register dynamic functions for a single field and its nested fields
func (fb *FormBuilder) registerFieldDynamicFunctions(field *Field, path string) {
	// Build the current field's path
	fieldPath := field.ID
	if path != "" {
		fieldPath = path + "." + field.ID
	}

	// Check if the field has dynamic options
	if field.Options != nil && field.Options.Type == OptionsTypeDynamic &&
		field.Options.DynamicSource != nil && field.Options.DynamicSource.Type == "function" {

		dynamicSource := field.Options.DynamicSource

		// If there's a DirectFunction, register it with a path-based name
		if dynamicSource.DirectFunction != nil {
			// Use the field ID as the function name (with path prefix)
			functionName := fieldPath

			// If there's already an explicitly set function name, use that instead
			if dynamicSource.FunctionName != "" &&
				!strings.HasPrefix(dynamicSource.FunctionName, "direct_func_") {
				functionName = dynamicSource.FunctionName
			} else {
				// Update the function name in the dynamic source
				dynamicSource.FunctionName = functionName
			}

			// Register the function with the schema
			fb.schema.RegisterFunction(functionName, dynamicSource.DirectFunction)
		}
	}

	// Process nested fields recursively
	if field.Nested != nil {
		for _, nestedField := range field.Nested {
			fb.registerFieldDynamicFunctions(nestedField, fieldPath)
		}
	}
}

File: react/core/form-builder.ts
import {
  APIKeyBuilder,
  BasicAuthBuilder,
  JWTBuilder,
  OAuth2Builder,
  SAMLBuilder,
} from "./auth-builder";
import { FieldBuilder } from "./field-builder";
import { FormSchemaImpl } from "./form-schema";
import {
  APIFieldBuilder,
  AnyOfFieldBuilder,
  ArrayFieldBuilder,
  AuthFieldBuilder,
  BranchFieldBuilder,
  CustomFieldBuilder,
  GroupFieldBuilder,
  OneOfFieldBuilder,
} from "./specialized-field-builder";
import { type AuthStrategy, type Field, FormType } from "./types";

/**
 * FormBuilder provides a fluent API for creating form schemas
 */
export class FormBuilder {
  private schema: FormSchemaImpl;

  /**
   * Create a new form builder
   * @param id Unique form identifier
   * @param title Form display title
   */
  constructor(id: string, title: string) {
    this.schema = new FormSchemaImpl(id, title);
  }

  /**
   * Creates a new authentication form builder
   * @param id Unique form identifier
   * @param title Form display title
   * @param authType Authentication strategy
   * @returns New form builder instance
   */
  static createAuthForm(
    id: string,
    title: string,
    authType: AuthStrategy,
  ): FormBuilder {
    const builder = new FormBuilder(id, title);
    builder.schema.type = FormType.Auth;
    builder.schema.authType = authType;
    return builder;
  }

  /**
   * Sets the form description
   * @param description Form description text
   */
  description(description: string): this {
    this.schema.description = description;
    return this;
  }

  /**
   * Sets the form type
   * @param formType Form type
   */
  formType(formType: FormType): this {
    this.schema.type = formType;
    return this;
  }

  /**
   * Sets the authentication type for auth forms
   * @param authType Authentication strategy
   */
  authType(authType: AuthStrategy): this {
    if (this.schema.type === FormType.Auth) {
      this.schema.authType = authType;
    }
    return this;
  }

  /**
   * Sets a custom property on the form
   * @param key Property key
   * @param value Property value
   */
  property(key: string, value: any): this {
    this.schema.properties[key] = value;
    return this;
  }

  /**
   * Adds a field to the form
   * @param field Field to add
   */
  addField(field: Field): this {
    this.schema.addField(field);
    return this;
  }

  /**
   * Adds multiple fields to the form
   * @param fields Fields to add
   */
  addFields(...fields: Field[]): this {
    this.schema.addFields(...fields);
    return this;
  }

  /**
   * Adds a text field to the form
   * @param id Field identifier
   * @param label Field display label
   */
  textField(id: string, label: string): FieldBuilder {
    const builder = new FieldBuilder(id, "text", label);
    this.addField(builder.build());
    return builder;
  }

  /**
   * Adds a textarea field to the form
   * @param id Field identifier
   * @param label Field display label
   */
  textareaField(id: string, label: string): FieldBuilder {
    const builder = new FieldBuilder(id, "textarea", label);
    this.addField(builder.build());
    return builder;
  }

  /**
   * Adds a number field to the form
   * @param id Field identifier
   * @param label Field display label
   */
  numberField(id: string, label: string): FieldBuilder {
    const builder = new FieldBuilder(id, "number", label);
    this.addField(builder.build());
    return builder;
  }

  /**
   * Adds an email field to the form
   * @param id Field identifier
   * @param label Field display label
   */
  emailField(id: string, label: string): FieldBuilder {
    const builder = new FieldBuilder(id, "email", label);
    this.addField(builder.build());
    return builder;
  }

  /**
   * Adds a password field to the form
   * @param id Field identifier
   * @param label Field display label
   */
  passwordField(id: string, label: string): FieldBuilder {
    const builder = new FieldBuilder(id, "password", label);
    this.addField(builder.build());
    return builder;
  }

  /**
   * Adds a select field to the form
   * @param id Field identifier
   * @param label Field display label
   */
  selectField(id: string, label: string): FieldBuilder {
    const builder = new FieldBuilder(id, "select", label);
    this.addField(builder.build());
    return builder;
  }

  /**
   * Adds a multi-select field to the form
   * @param id Field identifier
   * @param label Field display label
   */
  multiSelectField(id: string, label: string): FieldBuilder {
    const builder = new FieldBuilder(id, "multiselect", label);
    this.addField(builder.build());
    return builder;
  }

  /**
   * Adds a checkbox field to the form
   * @param id Field identifier
   * @param label Field display label
   */
  checkboxField(id: string, label: string): FieldBuilder {
    const builder = new FieldBuilder(id, "checkbox", label);
    this.addField(builder.build());
    return builder;
  }

  /**
   * Adds a radio button field to the form
   * @param id Field identifier
   * @param label Field display label
   */
  radioField(id: string, label: string): FieldBuilder {
    const builder = new FieldBuilder(id, "radio", label);
    this.addField(builder.build());
    return builder;
  }

  /**
   * Adds a date field to the form
   * @param id Field identifier
   * @param label Field display label
   */
  dateField(id: string, label: string): FieldBuilder {
    const builder = new FieldBuilder(id, "date", label);
    this.addField(builder.build());
    return builder;
  }

  /**
   * Adds a time field to the form
   * @param id Field identifier
   * @param label Field display label
   */
  timeField(id: string, label: string): FieldBuilder {
    const builder = new FieldBuilder(id, "time", label);
    this.addField(builder.build());
    return builder;
  }

  /**
   * Adds a datetime field to the form
   * @param id Field identifier
   * @param label Field display label
   */
  dateTimeField(id: string, label: string): FieldBuilder {
    const builder = new FieldBuilder(id, "datetime", label);
    this.addField(builder.build());
    return builder;
  }

  /**
   * Adds a file upload field to the form
   * @param id Field identifier
   * @param label Field display label
   */
  fileField(id: string, label: string): FieldBuilder {
    const builder = new FieldBuilder(id, "file", label);
    this.addField(builder.build());
    return builder;
  }

  /**
   * Adds an image upload field to the form
   * @param id Field identifier
   * @param label Field display label
   */
  imageField(id: string, label: string): FieldBuilder {
    const builder = new FieldBuilder(id, "image", label);
    this.addField(builder.build());
    return builder;
  }

  /**
   * Adds a switch field to the form
   * @param id Field identifier
   * @param label Field display label
   */
  switchField(id: string, label: string): FieldBuilder {
    const builder = new FieldBuilder(id, "switch", label);
    this.addField(builder.build());
    return builder;
  }

  /**
   * Adds a slider field to the form
   * @param id Field identifier
   * @param label Field display label
   */
  sliderField(id: string, label: string): FieldBuilder {
    const builder = new FieldBuilder(id, "slider", label);
    this.addField(builder.build());
    return builder;
  }

  /**
   * Adds a rating field to the form
   * @param id Field identifier
   * @param label Field display label
   */
  ratingField(id: string, label: string): FieldBuilder {
    const builder = new FieldBuilder(id, "rating", label);
    this.addField(builder.build());
    return builder;
  }

  /**
   * Adds a color picker field to the form
   * @param id Field identifier
   * @param label Field display label
   */
  colorField(id: string, label: string): FieldBuilder {
    const builder = new FieldBuilder(id, "color", label);
    this.addField(builder.build());
    return builder;
  }

  /**
   * Adds a hidden field to the form
   * @param id Field identifier
   * @param value Field value
   */
  hiddenField(id: string, value: any): FieldBuilder {
    const builder = new FieldBuilder(id, "hidden", "");
    builder.defaultValue(value);
    this.addField(builder.build());
    return builder;
  }

  /**
   * Adds a rich text editor field to the form
   * @param id Field identifier
   * @param label Field display label
   */
  richTextField(id: string, label: string): FieldBuilder {
    const builder = new FieldBuilder(id, "richtext", label);
    this.addField(builder.build());
    return builder;
  }

  /**
   * Adds a section separator to the form
   * @param id Field identifier
   * @param label Field display label
   */
  sectionField(id: string, label: string): FieldBuilder {
    const builder = new FieldBuilder(id, "section", label);
    this.addField(builder.build());
    return builder;
  }

  /**
   * Adds a group field to the form
   * @param id Field identifier
   * @param label Field display label
   */
  groupField(id: string, label: string): GroupFieldBuilder {
    const builder = new GroupFieldBuilder(id, label);
    this.addField(builder.build());
    return builder;
  }

  /**
   * Adds an array field to the form
   * @param id Field identifier
   * @param label Field display label
   */
  arrayField(id: string, label: string): ArrayFieldBuilder {
    const builder = new ArrayFieldBuilder(id, label);
    this.addField(builder.build());
    return builder;
  }

  /**
   * Adds a oneOf field to the form
   * @param id Field identifier
   * @param label Field display label
   */
  oneOfField(id: string, label: string): OneOfFieldBuilder {
    const builder = new OneOfFieldBuilder(id, label);
    this.addField(builder.build());
    return builder;
  }

  /**
   * Adds an anyOf field to the form
   * @param id Field identifier
   * @param label Field display label
   */
  anyOfField(id: string, label: string): AnyOfFieldBuilder {
    const builder = new AnyOfFieldBuilder(id, label);
    this.addField(builder.build());
    return builder;
  }

  /**
   * Adds an API integration field to the form
   * @param id Field identifier
   * @param label Field display label
   */
  apiField(id: string, label: string): APIFieldBuilder {
    const builder = new APIFieldBuilder(id, label);
    this.addField(builder.build());
    return builder;
  }

  /**
   * Adds an authentication field to the form
   * @param id Field identifier
   * @param label Field display label
   */
  authField(id: string, label: string): AuthFieldBuilder {
    const builder = new AuthFieldBuilder(id, label);
    this.addField(builder.build());
    return builder;
  }

  /**
   * Adds an OAuth authentication field to the form
   * @param id Field identifier
   * @param label Field display label
   */
  oAuthField(id: string, label: string): OAuth2Builder {
    const builder = new OAuth2Builder(id, label);
    this.addField(builder.build());
    return builder;
  }

  /**
   * Adds a basic authentication field to the form
   * @param id Field identifier
   * @param label Field display label
   */
  basicAuthField(id: string, label: string): BasicAuthBuilder {
    const builder = new BasicAuthBuilder(id, label);
    this.addField(builder.build());
    return builder;
  }

  /**
   * Adds an API key authentication field to the form
   * @param id Field identifier
   * @param label Field display label
   */
  apiKeyField(id: string, label: string): APIKeyBuilder {
    const builder = new APIKeyBuilder(id, label);
    this.addField(builder.build());
    return builder;
  }

  /**
   * Adds a JWT authentication field to the form
   * @param id Field identifier
   * @param label Field display label
   */
  jwtField(id: string, label: string): JWTBuilder {
    const builder = new JWTBuilder(id, label);
    this.addField(builder.build());
    return builder;
  }

  /**
   * Adds a SAML authentication field to the form
   * @param id Field identifier
   * @param label Field display label
   */
  samlField(id: string, label: string): SAMLBuilder {
    const builder = new SAMLBuilder(id, label);
    this.addField(builder.build());
    return builder;
  }

  /**
   * Adds a workflow branch field to the form
   * @param id Field identifier
   * @param label Field display label
   */
  branchField(id: string, label: string): BranchFieldBuilder {
    const builder = new BranchFieldBuilder(id, label);
    this.addField(builder.build());
    return builder;
  }

  /**
   * Adds a custom field to the form
   * @param id Field identifier
   * @param label Field display label
   */
  customField(id: string, label: string): CustomFieldBuilder {
    const builder = new CustomFieldBuilder(id, label);
    this.addField(builder.build());
    return builder;
  }

  /**
   * Finalizes and returns the form schema
   * @returns The completed form schema
   */
  build(): FormSchemaImpl {
    // Sort fields by order before returning
    this.schema.sortFields();
    return this.schema;
  }
}
